apps/studio/components/interfaces/Storage/EditBucketModal.tsx

```diff
@@ -16,256 +42,332 @@ import { useBucketUpdateMutation } from 'data/storage/bucket-update-mutation'
-      errors.formatted_size_limit = 'File size upload limit has to be at least 0'
-    }
-    return errors
-  }
+  const form = useForm<z.infer<typeof BucketSchema>>({
+    resolver: zodResolver(BucketSchema),
+    defaultValues: {
+      name: bucket?.name ?? '',
+      public: bucket?.public,
+      has_file_size_limit: isNonNullable(bucket?.file_size_limit),
+      formatted_size_limit: fileSizeLimit ?? 0,
+      allowed_mime_types: (bucket?.allowed_mime_types ?? []).join(', '),
+    },
+    values: {
+      name: bucket?.name ?? '',
+      public: bucket?.public,
+      has_file_size_limit: isNonNullable(bucket?.file_size_limit),
+      formatted_size_limit: fileSizeLimit ?? 0,
+      allowed_mime_types: (bucket?.allowed_mime_types ?? []).join(', '),
+    },
+    mode: 'onSubmit',
+  })
+
+  const isPublicBucket = form.watch('public')
+  const hasFileSizeLimit = form.watch('has_file_size_limit')
+  const formattedSizeLimit = form.watch('formatted_size_limit')
+  const isChangingBucketVisibility = bucket?.public !== isPublicBucket
+  const isMakingBucketPrivate = bucket?.public && !isPublicBucket
+  const isMakingBucketPublic = !bucket?.public && isPublicBucket
 
-  const onSubmit = async (values: any) => {
+  const onSubmit: SubmitHandler<z.infer<typeof BucketSchema>> = async (values) => {
     if (bucket === undefined) return console.error('Bucket is required')
     if (ref === undefined) return console.error('Project ref is required')
 
-    updateBucket({
-      projectRef: ref,
-      id: bucket.id,
-      isPublic: values.public,
-      file_size_limit: values.has_file_size_limit
-        ? convertToBytes(values.formatted_size_limit, selectedUnit)
-        : null,
-      allowed_mime_types:
-        values.allowed_mime_types.length > 0
-          ? values.allowed_mime_types.split(',').map((x: string) => x.trim())
+    updateBucket(
+      {
+        projectRef: ref,
+        id: bucket.id,
+        isPublic: values.public,
+        file_size_limit: values.has_file_size_limit
+          ? convertToBytes(values.formatted_size_limit, selectedUnit as StorageSizeUnits)
           : null,
-    })
+        allowed_mime_types:
+          values.allowed_mime_types.length > 0
+            ? values.allowed_mime_types.split(',').map((x: string) => x.trim())
+            : null,
+      },
+      {
+        onSuccess: () => {
+          toast.success(`Successfully updated bucket "${bucket?.name}"`)
+          onClose()
+        },
+      }
+    )
   }
 
-  useEffect(() => {
-    if (visible) {
-      const { unit } = convertFromBytes(bucket?.file_size_limit ?? 0)
-      setSelectedUnit(unit)
-      setShowConfiguration(false)
-    }
-  }, [visible])
-
   return (
-    <Modal
-      hideFooter
-      visible={visible}
-      size="medium"
-      header={`Edit bucket "${bucket?.name}"`}
-      onCancel={onClose}
+    <Dialog
+      open
+      onOpenChange={(open) => {
+        if (!open) {
+          form.reset()
```

Comment:
thinking about it, `form.reset()` might not be necessary at all because the moment the dialog closes its unmounted from the component tree, which should reset its state implicitly

might come back and update this after testing that theory